modules/tinymce/src/core/main/ts/api/ThemeManager.ts

```diff
@@ -16,7 +16,7 @@ export interface Theme {
   execCommand?: (command: string, ui?: boolean, value?: any) => boolean;
   destroy?: () => void;
   init?: (editor: Editor, url: string) => void;
-  renderUI?: () => RenderResult;
+  renderUI?: () => Promise<RenderResult> | RenderResult;
```

Comment:
Perhaps I wasn't clear before. I'm not changing my mind about this without a much wider discussion. It's breaking a public API which we can't do until a major release.

Our `tinymce.d.ts` currently contains:
```
interface Theme {
    ui?: any;
    inline?: any;
    execCommand?: (command: string, ui?: boolean, value?: any) => boolean;
    destroy?: () => void;
    init?: (editor: Editor, url: string) => void;
    renderUI?: () => {
        iframeContainer?: HTMLIFrameElement;
        editorContainer: HTMLElement;
        api?: Partial<EditorUiApi>;
    };
    getNotificationManagerImpl?: () => NotificationManagerImpl;
    getWindowManagerImpl?: () => WindowManagerImpl;
}
```

And after this change:
```
interface RenderResult {
    iframeContainer?: HTMLElement;
    editorContainer: HTMLElement;
    api?: Partial<EditorUiApi>;
}
interface Theme {
    ui?: any;
    inline?: any;
    execCommand?: (command: string, ui?: boolean, value?: any) => boolean;
    destroy?: () => void;
    init?: (editor: Editor, url: string) => void;
    renderUI?: () => Promise<RenderResult> | RenderResult;
    getNotificationManagerImpl?: () => NotificationManagerImpl;
    getWindowManagerImpl?: () => WindowManagerImpl;
}
```

Will this break any customer integrations? Almost certainly not. I would be very surprised if _anyone_ is relying on the old type. But it's still a breaking change under the strict semver rules.

Reply:
Should we risk it here since it needs to be a customer who uses typescript and has custom themes that can't be that many people. Also it will not break cloud since both sync and async would work.

But if they have some predicate that acts on the output of the renderUI and expecting that to be only a RenderResult then it will fail to compile. But it would be fairly obvious how to fix that.

Reply:
I think we are fine we are breaking types every minor release it seems. Sometimes we go from 'something' to 'something | null' etc in public apis that is what was breaking the `casechange` plugin for example that was bumped from 6.0.0 to 6.3.1 and it seems that we are doing that quite often not seen any complaints so I think it's ok to do this change here it's even less risky than the other things we broken before in other minors since it's a fringe api.

Reply:
At the end of the day semver is best described as an agreement, not an enforceable rule. The definition of what constitutes a breaking change is also debatable.

If we've decided to relax a bit based on what we think will actually break customer integrations that's fine, but since December 2019 we've practiced strict semver. I just wanted to point out that this doesn't fit that definition.